File: backend/src/models/Category.js
const mongoose = require('mongoose');

const categorySchema = new mongoose.Schema({
  name: {
    type: String,
    required: true,
    trim: true,
    unique: true
  },
  description: {
    type: String,
    required: false,
    trim: true,
    default: ''
  },
  icon: {
    type: String,
    required: false,
    default: 'school'
  },
  subcategories: {
    type: [String],
    default: []
  },
  isPopular: {
    type: Boolean,
    default: false
  },
  isFeatured: {
    type: Boolean,
    default: false
  },
  level: {
    type: Number,
    enum: [1, 2, 3, 4, 5], // 1 = All levels, 2 = Fluency, 3 = In-demand, 4 = Career-ready, 5 = Growth
    default: 1
  },
  courses: [{
    type: mongoose.Schema.Types.ObjectId,
    ref: 'Course'
  }],
  createdAt: {
    type: Date,
    default: Date.now
  },
  updatedAt: {
    type: Date,
    default: Date.now
  }
});

// Index for search optimization
categorySchema.index({ name: 'text', description: 'text' });

// Performance indexes for common queries
categorySchema.index({ level: 1, name: 1 }); // For getCategoriesByLevel
categorySchema.index({ isPopular: 1, name: 1 }); // For getPopularCategories
categorySchema.index({ isFeatured: 1, name: 1 }); // For getFeaturedCategories
categorySchema.index({ level: 1, isPopular: 1, isFeatured: 1 }); // Compound index for filtering

// Update the updatedAt field before saving
categorySchema.pre('save', function() {
  this.updatedAt = Date.now();
});

module.exports = mongoose.model('Category', categorySchema);

File: frontend/lib/models/category.dart
import 'dart:convert';

class Category {
  final String id;
  final String name;
  final String description;
  final String icon;
  final List<String> subcategories;
  final bool isPopular;
  final bool isFeatured;
  final int level;

  Category({
    required this.id,
    required this.name,
    required this.description,
    required this.icon,
    required this.subcategories,
    this.isPopular = false,
    this.isFeatured = false,
    this.level = 1,
  });

  factory Category.fromJson(Map<String, dynamic> json) {
    return Category(
      id: json['id'] as String? ?? json['_id']?.toString() ?? '',
      name: json['name'] as String? ?? '',
      description: json['description'] as String? ?? '',
      icon: json['icon'] as String? ?? '',
      subcategories: json['subcategories'] is List
          ? List<String>.from(json['subcategories'])
          : [],
      isPopular: json['isPopular'] as bool? ?? false,
      isFeatured: json['isFeatured'] as bool? ?? false,
      level: json['level'] as int? ?? 1,
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'name': name,
      'description': description,
      'icon': icon,
      'subcategories': subcategories,
      'isPopular': isPopular,
      'isFeatured': isFeatured,
      'level': level,
    };
  }

  @override
  String toString() {
    return 'Category(id: $id, name: $name, level: $level)';
  }

  @override
  bool operator ==(Object other) =>
      identical(this, other) ||
      other is Category &&
          runtimeType == other.runtimeType &&
          id == other.id;

  @override
  int get hashCode => id.hashCode;
}
